timemachine/cpp/src/bound_potential.cu

```diff
@@ -37,14 +37,30 @@ void BoundPotential::execute_host(
     gpuErrchk(cudaMemset(d_du_dl.data, 0, d_du_dl.size));
     gpuErrchk(cudaMemset(d_u.data, 0, d_u.size));
 
-    this->execute_device(
-        N, d_x.data, d_box.data, lambda, d_du_dx.data, nullptr, d_du_dl.data, d_u.data, static_cast<cudaStream_t>(0));
-
+    cudaStream_t stream = static_cast<cudaStream_t>(0);
+    this->execute_device(N, d_x.data, d_box.data, lambda, d_du_dx.data, nullptr, d_du_dl.data, d_u.data, stream);
+    gpuErrchk(cudaStreamSynchronize(stream));
     d_du_dx.copy_to(h_du_dx);
     d_du_dl.copy_to(h_du_dl);
     d_u.copy_to(h_u);
 };
 
+void BoundPotential::set_params_device(
+    const std::vector<int> device_shape, const double *d_new_params, const cudaStream_t stream) {
+    int updated_size = 1;
+    for (auto s : device_shape) {
+        updated_size *= s;
+    }
+    if (updated_size > 0) {
+        if (updated_size > this->size()) {
+            d_p.reset(new DeviceBuffer<double>(updated_size));
+        }
+        gpuErrchk(cudaMemcpyAsync(
+            d_p->data, d_new_params, updated_size * sizeof(*d_p->data), cudaMemcpyDeviceToDevice, stream));
```

Comment:
If the argument is changed to a DeviceBuffer, this could be
```suggestion
            d_p->data, new_params.data, new_params.size, cudaMemcpyDeviceToDevice, stream));
```